File: Queue/Himadri/lru-cache.cc
//solution by Shefali Sonker

// { Driver Code Starts

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int key;
    int value;
    Node *next,*pre;
    Node(int key, int value)
    {
        this->key=key;
        this->value=value;
        next=pre=NULL;
    }

};


 // } Driver Code Ends



/*
struct Node
{
    int key;
    int value;
    Node *next,*pre;
    Node(int key, int value)
    {
        this->key=key;
        this->value=value;
        next=pre=NULL;
    }

};
*/

class LRUCache
{
    private:

    static unordered_map<int,Node *> hsmap;
    static int capacity,count;
    static Node *head,*tail;

    public:

    LRUCache(int cap);
    int get(int key);
    void set(int key, int value);


};

int cachesize = 0;
map <int, int> mp;
deque<int> dq;

LRUCache::LRUCache(int cap) {
    cachesize = cap;
    dq.clear();
    mp.clear();
}

void LRUCache::set(int key, int value){
    if(mp.find(key)==mp.end()){
        mp[key] = value;
        if(dq.size() < cachesize) dq.push_back(key);
        else {
            int temp = dq.front();
            dq.pop_front();
            mp.erase(temp);
            dq.push_back(key);
        }
    }
    else{
        mp[key] = value;
        for(deque<int>::iterator itr = dq.begin() ; itr!=dq.end(); itr++) {
            if(*itr==key){
                dq.erase(itr);
                break;
            }
        }
        dq.push_back(key);
    }
    return;
}

int LRUCache::get(int key){
    if(mp.find(key) == mp.end()) return -1;
    set(key, mp[key]);
    return mp[key];
}




// { Driver Code Starts.

//Initializing static members
unordered_map<int, Node*> temp;
int LRUCache::capacity=0;
Node * LRUCache::head= new Node(0,0);
Node *  LRUCache::tail= new Node(0,0);
int LRUCache::count= 0;
unordered_map<int, Node*> LRUCache::hsmap= temp;



int main() {
	int t;
	cin>>t;
	while(t--)
	{

	    int capacity;
	    cin>>capacity;
	    int queries;
	    cin>>queries;

	    LRUCache *cache = new LRUCache(capacity);
	    while(queries--)
	    {
	        string q;
	        cin>>q;
	        if(q=="SET")
	        {
	            int key;
	            cin>>key;
	            int value;
	            cin>>value;

	            cache->set(key,value);

	        }
	        else
	        {
	            int key;
	            cin>>key;

	            cout<<cache->get(key)<<" ";

	        }
	    }

	    cout<<endl;
	}
	return 0;
}  // } Driver Code Ends

File: Queue/Harshit/LRU_cache.java
// { Driver Code Starts

import java.io.*;
import java.util.*;
import java.lang.*;


class Node
{
    int key, value;
    Node next, pre;

    Node(int key, int value)
    {
        this.key = key;
        this.value = value;
        next = pre = null;
    }
}

public class LRUDesign {

    public static void main(String[] args)throws IOException {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(read.readLine());

        while(t-- > 0)
        {

            int capacity = Integer.parseInt(read.readLine());
            int queries = Integer.parseInt(read.readLine());
            LRUCache cache = new LRUCache(capacity);
            String str[] = read.readLine().trim().split(" ");
            int len = str.length;
            int itr = 0;

            for(int i = 0; (i < queries) && (itr < len); i++)
            {
                String queryType = str[itr++];
                if(queryType.equals("SET"))
                {
                    int key = Integer.parseInt(str[itr++]);
                    int value = Integer.parseInt(str[itr++]);
                    cache.set(key, value);
                }
                if(queryType.equals("GET")){
                    int key = Integer.parseInt(str[itr++]);
                    System.out.print(cache.get(key) + " ");
                }

            }
            System.out.println();

        }
    }
}

// } Driver Code Ends

/*
class Node
{
    int key, value;
    Node next, pre;

    Node(int key, int value)
    {
        this.key = key;
        this.value = value;
        next = pre = null;
    }
}
*/

class LRUCache
{
    private static HashMap<Integer, Node> map;
    private static int capacity, count;
    private static Node head, tail;

    LRUCache(int cap)
    {
        capacity=cap;
        
        map=new HashMap<>();
        capacity=cap;
        count=0;
        head=new Node(0,0);
        tail=new Node(0,0);
        head.next=tail;
        tail.pre=head;
        head.pre=null;
        tail.next=null;
        
        // Intialize the cache capacity with the given 
        // cap
    }
    
    public static void addToHead(Node node)
    {
         node.next = head.next; 
        node.next.pre = node; 
        node.pre = head; 
        head.next = node; 
        
    }
    
    public static void delNode(Node node)
    {
        node.next.pre=node.pre;
        node.pre.next=node.next;
        
        
    }
    

    // This method works in O(1)
    public static int get(int key)
    {
        if(map.containsKey(key))
        {
            Node node=map.get(key);
            int r=node.value;
            delNode(node);
            addToHead(node);
            return r;
    
        }
        return -1;
        
        
        // your code here
    }

    // This method works in O(1)
    public static void set(int key, int value)
    {
        if(map.containsKey(key))
        {
            Node node=map.get(key);
            node.value=value;
            delNode(node);
            addToHead(node);
            // map.replace(key,node);
           
        }
        else{
            
            Node node=new Node(key,value);
            map.put(key,node);
          
        if(count<capacity)
        {
            addToHead(node);
            count++;
            
        }
        else{
            map.remove(tail.pre.key);
            delNode(tail.pre);
            addToHead(node);
            
        }
            
            
        }
        
        // count++;
        
        
        
        
        
        
        
        
        
        
        
        // your code here
    }
}

